master/buildbot/db/builds.py

```diff
@@ -276,6 +276,84 @@ def thd(conn):
+        :param buildername: an string value with builder name, this name must exists in master.cfg
+        :param slavepool: a string value with slave pool name
+        :param number: an integer value with build number. Must be unique with build request id
+        :param slavename: a string value with slave name
+        :param results: an integer value with results status. See available options: master.buildbot.status.results
+        :param codebase: a string value with codebase of repository
+        :return: defer value
+        """
+        def thd(conn):
+            transaction = conn.begin()
+            try:
+                # Create sourcestampsets
+                r = conn.execute(self.db.model.sourcestampsets.insert(), dict())
+                sourcestampsset_id = r.inserted_primary_key[0]
+
+                # Create sourcestamps
+                conn.execute(self.db.model.sourcestamps.insert(), {
+                    'branch': branch,
+                    'revision': revision,
+                    'patchid': None,
+                    'repository': repository,
+                    'codebase': codebase,
+                    'project': project,
+                    'sourcestampsetid': sourcestampsset_id,
+                })
+
+                # Create buildsets
+                res = conn.execute(self.db.model.buildsets.insert(), {
+                    'reason': reason,
+                    'sourcestampsetid': sourcestampsset_id,
+                    'submitted_at': submitted_at,
+                    'complete': bool(complete_at),
+                    'complete_at': complete_at,
+                    'results': results,
+                })
+                buildset_id = res.inserted_primary_key[0]
+
+                # Create buildrequests
+                res = conn.execute(self.db.model.buildrequests.insert(), {
+                    'buildsetid': buildset_id,
+                    'buildername': buildername,
+                    'proiority': 50,
+                    'complete': bool(complete_at),
+                    'results': results,
+                    'submitted_at': submitted_at,
+                    'complete_at': complete_at,
+                    'slavepool': slavepool,
+                })
+                buildrequest_id = res.inserted_primary_key[0]
+
+                # Create builds
+                conn.execute(self.db.model.builds.insert(), {
+                    'number': number,
+                    'brid': buildrequest_id,
+                    'slavename': slavename,
+                    'start_time': submitted_at,
+                    'finish_time': complete_at,
+                })
+                transaction.commit()
+                return True
+            except Exception as e:
+                print("Exception occurs during create new build", e)
+                transaction.rollback()
+                return False
```

Comment:
`return True | False` looks weird to me in Python.

Why not just re-raise the exception after rollback?

Reply:
Solved 👍